core/controller/src/main/scala/whisk/core/controller/actions/SequenceActions.scala

```diff
@@ -247,40 +245,47 @@ protected[actions] trait SequenceActions {
     //
     // This action/parameter resolution is done in futures; the execution starts as soon as the first component
     // is resolved.
-    val resolvedFutureActions = resolveDefaultNamespace(components, user) map { c =>
+    val resolvedFutureActions: Seq[Future[WhiskActionMetaData]] = resolveDefaultNamespace(components, user) map { c =>
       WhiskActionMetaData.resolveActionAndMergeParameters(entityStore, c)
     }
 
     // this holds the initial value of the accounting structure, including the input boxed as an ActivationResponse
-    val initialAccounting = Future.successful {
-      SequenceAccounting(atomicActionCnt, ActivationResponse.payloadPlaceholder(inputPayload))
-    }
-
-    // execute the actions in sequential blocking fashion
-    resolvedFutureActions
-      .foldLeft(initialAccounting) { (accountingFuture, futureAction) =>
-        accountingFuture.flatMap { accounting =>
-          if (accounting.atomicActionCnt < actionSequenceLimit) {
-            invokeNextAction(user, futureAction, accounting, cause).flatMap { accounting =>
-              if (!accounting.shortcircuit) {
-                Future.successful(accounting)
-              } else {
-                // this is to short circuit the fold
-                Future.failed(FailedSequenceActivation(accounting)) // terminates the fold
-              }
+    val initialAccounting = SequenceAccounting(atomicActionCnt, ActivationResponse.payloadPlaceholder(inputPayload))
+
+    Future
+      .sequence(resolvedFutureActions)
+      .flatMap(actions =>
```

Comment:
Gave refactoring a shot. Turned `resolvedFutureActions` into a `Future[List[resolvedFutureActions]]`. My understanding is that the `flapMap` block won't be executed if one or more of the actions does not exist and instead execution will jump to the second `recoverWith` on line 277.

Reply:
@rabbah, what do you think?

Reply:
i think this is safer but now we wait for all the futures to resolve on the fetch - and don't pipeline any of the invokes, so worse on performance for a long sequence.

i think we need
1. additional test to cover missing package
2. additional test to cover missing bindings
3. a test with 2 sequence components, where the first is fetched but the second fails, the accounting object should show the execution that accounts for the first component's execution.

1 & 2 to make sure all no doc exceptions are handles consistently.
3 will require refactoring the code in the function so that you can test the independent components without having to mess with the data store. This is what I meant by refactoring.